### ISSUE TITLE
Incorrect error message in condition/action method validation

### ISSUE THREAD

**Issue by khaledabderrahim** [j-easy/easy-rules#165]
the message of the exception says :

```
Condition method  defined in rule must be public, may have parameters annotated with @Fact (and/or exactly one parameter of type or extending Facts)
```

but in the implementation :-1:  just if there is no parameters annotated with `@Fact`
then the only parameter must be subtype of `Facts`

```java
@Rule
public class MyRule {
	
	@Action
	public void myAction( @Fact("i") Integer i,Float f ) {
		System.out.println("action method ");
	}
	@Condition
	 public boolean myCondition() {
		return true;
	}
	
	@Priority
	int myRulePriority() {
		return 0;		
	}
}
```

```java
Method m = myRule.getClass().getDeclaredMethod("myAction", Integer.class,Float.class);
			System.out.println(isActionMethodWellDefined(m));
```
will print true 
solution :   change 

```java
if(parameterTypes.length == 1 && notAnnotatedParameterCount == 1){
            return Facts.class.isAssignableFrom(parameterTypes[0]);
        }
```
to :

```java
if(parameterTypes.length == 1 || notAnnotatedParameterCount == 1){
            return Facts.class.isAssignableFrom(parameterTypes[0]);
        }
```

**Comment by benas** [comment]
The message of the exception you mentioned is about the condition method, but your example is about the action method. I'm not sure I understand your point.

Can you please clarify? The best way is to provide a test with what you are expecting so I can understand the problem. Thank you upfront.

**Comment by benas** [comment]
Hi @khaledabderrahim ,

Thank you for opening this issue! After further analysis, this is indeed a bug in Easy Rules. It is more than the error message being incorrect (and confusing!), it is actually a rule validation issue. The example of rule you gave is invalid (parameter `f` should either be annotated with `@Fact` or be of type `Facts`), but is considered valid. Here is a failing test with 3.2.0:

```java
@Test(expected = IllegalArgumentException.class)
public void methodParametersShouldAllBeAnnotatedWithFactUnlessExactlyOneOfThemIsOfTypeFacts() {

    @Rule
    class MyRule {

        @Action
        public void myAction( @Fact("i") Integer i,Float f ) {
            System.out.println("action method ");
        }
        @Condition
        public boolean myCondition() {
            return true;
        }

        @Priority
        int myRulePriority() {
            return 0;
        }
    }

    new RuleDefinitionValidator().validateRuleDefinition(new MyRule());
}
```

 is how to import the snapshot version if you don't know how to do it). 

> solution : change

The solution you suggest breaks several tests, so I fixed the issue differently (See df6e0f7b5628fc8cce6c9bf94702ed9cfb0d1a8e). I also updated the error message to be more explicit.

The fix has been deployed in version `3.3.0-SNAPSHOT`. Can you give it a try and tell me if this is ok for you? [Here](https://github.com/j-easy/easy-rules/wiki/getting-started#use-a-snapshot-version) is how import the snapshot version if you don't know how to do it.

Kr,
Mahmoud
